src/workerd/api/streams/writable.c++

```diff
@@ -280,6 +282,137 @@ jsg::Ref<WritableStream> WritableStream::constructor(
+class WritableStreamRpcAdapter final: public capnp::ExplicitEndOutputStream {
+public:
+  WritableStreamRpcAdapter(kj::Own<WritableStreamSink> inner)
+      : inner(kj::mv(inner)) {}
+  ~WritableStreamRpcAdapter() noexcept(false) {
+    weakRef->invalidate();
+    doneFulfiller->fulfill();
+  }
+
+  // Returns a promise that resolves when the stream is dropped. If the promise is canceled before
+  // that, the stream is revoked.
+  kj::Promise<void> waitForCompletionOrRevoke() {
+    auto paf = kj::newPromiseAndFulfiller<void>();
+    doneFulfiller = kj::mv(paf.fulfiller);
+
+    return paf.promise.attach(kj::defer([weakRef = weakRef->addRef()]() mutable {
+      KJ_IF_SOME(obj, weakRef->tryGet()) {
+        // Stream is still alive, revoke it.
+        if (!obj.canceler.isEmpty()) {
+          obj.canceler.cancel(cancellationException());
+        }
+        obj.inner = kj::none;
+      }
+    }));
+  }
+
+  kj::Promise<void> write(const void* buffer, size_t size) override {
+    return canceler.wrap(getInner().write(buffer, size));
+  }
+  kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) override {
+    return canceler.wrap(getInner().write(pieces));
+  }
+
+  // TODO(perf): We can't properly implement tryPumpFrom(), which means that Cap'n Proto will
+  //   be unable to perform path shortening if the underlying stream turns out to be another capnp
+  //   stream. This isn't a huge deal, but might be nice to enable someday. It may require
+  //   significant refactoring of streams.
+
+  kj::Promise<void> whenWriteDisconnected() override {
+    // TODO(someday): WritableStreamSink doesn't give us a way to implement this.
+    return kj::NEVER_DONE;
+  }
+
+  kj::Promise<void> end() override {
+    return canceler.wrap(getInner().end());
+  }
+
+private:
+  kj::Maybe<kj::Own<WritableStreamSink>> inner;
+  kj::Canceler canceler;
+  kj::Own<kj::PromiseFulfiller<void>> doneFulfiller;
+  kj::Own<WeakRef<WritableStreamRpcAdapter>> weakRef =
+      kj::refcounted<WeakRef<WritableStreamRpcAdapter>>(
+          kj::Badge<WritableStreamRpcAdapter>(), *this);
+
+  WritableStreamSink& getInner() {
+    return *KJ_UNWRAP_OR(inner, {
+      kj::throwFatalException(cancellationException());
+    });
+  }
+
+  static kj::Exception cancellationException() {
+    return JSG_KJ_EXCEPTION(DISCONNECTED, Error,
+        "WritableStream received over RPC was disconnected because the remote execution context "
+        "has endeded.");
+  }
+};
+
+}  // namespace
+
+void WritableStream::serialize(jsg::Lock& js, jsg::Serializer& serializer) {
+  // Serialize by effectively creating a `JsRpcStub` around this object and serializing that.
+  // Except we don't actually want to do _exactly_ that, because we do not want to actually create
+  // a `JsRpcStub` locally. So do the important parts of `JsRpcStub::constructor()` followed by
+  // `JsRpcStub::serialize()`.
```

Comment:
Likewise here, this should check to make sure the `WritableStream` is not locked. If it is, this should throw before doing anything else here.

Reply:
ff1c6fed